chpl/chpl-service/src/main/java/gov/healthit/chpl/upload/listing/ListingUploadManager.java

```diff
@@ -0,0 +1,377 @@
+        CertifiedProductSearchDetails listing =
+                listingDetailsHandler.parseAsListing(headingRecord, allListingRecords);
+        listing.setId(id);
+        listingNormalizer.normalize(listing);
+        return listing;
+    }
+
+    @Transactional
+    @PreAuthorize("@permissions.hasAccess(T(gov.healthit.chpl.permissions.Permissions).LISTING_UPLOAD, "
+            + "T(gov.healthit.chpl.permissions.domains.ListingUploadDomainPerissions).DELETE, #id)")
+    public void delete(Long id)
+            throws EntityRetrievalException, EntityNotFoundException, EntityCreationException, AccessDeniedException,
+            JsonProcessingException, ObjectMissingValidationException {
+        if (isListingUploadAvailableForDelete(id)) {
+            ListingUpload listingUploadBeforeDelete = listingUploadDao.getById(id);
+            listingUploadDao.delete(id);
+            String activityMsg = "Listing upload " + listingUploadBeforeDelete.getChplProductNumber() + " has been rejected.";
+            activityManager.addActivity(ActivityConcept.LISTING_UPLOAD, listingUploadBeforeDelete.getId(),
+                    activityMsg, listingUploadBeforeDelete, null);
+        }
+    }
+
+    private boolean isListingUploadAvailableForDelete(Long id)
+            throws EntityRetrievalException, ObjectMissingValidationException {
+
+        ListingUploadEntity entity = listingUploadDao.getEntityByIdIncludingDeleted(id);
+        if (entity.getDeleted()) {
+            ObjectMissingValidationException alreadyDeletedEx = new ObjectMissingValidationException();
+            alreadyDeletedEx.getErrorMessages()
+                    .add("This pending certified product has already been confirmed or rejected by another user.");
+            alreadyDeletedEx.setObjectId(entity.getChplProductNumber());
+            try {
+                UserDTO lastModifiedUserDto = userDao.getById(entity.getLastModifiedUser());
+                if (lastModifiedUserDto != null) {
+                    User lastModifiedUser = new User(lastModifiedUserDto);
+                    alreadyDeletedEx.setUser(lastModifiedUser);
+                } else {
+                    alreadyDeletedEx.setUser(null);
+                }
+            } catch (final UserRetrievalException ex) {
+                alreadyDeletedEx.setUser(null);
+            }
+            throw alreadyDeletedEx;
+        } else {
+            return true;
+        }
+
+    }
+
+    private void checkRequiredHeadings(CSVRecord headingRecord) throws ValidationException {
+        List<String> missingRequiredHeadings = Headings.getRequiredHeadings().stream()
+            .filter(heading -> !uploadUtil.hasHeading(heading, headingRecord))
+            .map(headingVal -> headingVal.getNamesAsString())
+            .collect(Collectors.toList());
+        if (missingRequiredHeadings != null && missingRequiredHeadings.size() > 0) {
```

Comment:
I don't think that Collecters.toList() will ever return NULL, so I'm not sure the null check is necessary.  No big deal.